### PR TITLE
Feature/validate placements

### PR DESCRIPTION
**Pull request by suzkiee**
## What does this change?

validate_placement method

## What does this fix?

Adds first conditional statement that checks if the ship length is equal to the array of coordinates in the argument. Also checks that the coordinates are consecutive horizontally on the board. 

## Is this a bug fix or a feature? 

feature 

## How was this tested? 
Tests were made for the first two conditional statements in the method.

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -21,8 +21,68 @@ def initialize
                "D4" => Cell.new("D4")
               }
   end
+  # looping through - nested loop for expansion
```
Comment (mkrumholz) [inline lib/board.rb:24]: save for another time.

File: lib/board.rb [hunk 2]
```diff
@@ -21,8 +21,68 @@ def initialize
...
+
+  def all_same?(array)
+    array.uniq.length == 1 
+  end
+
+  def not_all_same?(array)
+    array.uniq.length > 1 
+  end
+
+  def valid_placement?(ship, cells)
```
Comment (mkrumholz) [inline lib/board.rb:38]: yes! I think we should convert the letters to numbers/ords here and maybe make this its own helper method

File: lib/board.rb [hunk 3]
```diff
@@ -21,8 +21,68 @@ def initialize
...
+    letters = cells.map do |cell| 
+      cell.split('').first
+    end
+    #should we convert the letters to integers here instead of at the end?
+
+    numbers = cells.map do |cell|
+      cell.split('').last.to_i
+    end
+    # Could this be a case expression?
```
Comment (mkrumholz) [inline lib/board.rb:47]: Nope! We can remove this comment

File: lib/board.rb [hunk 4]
```diff
@@ -21,8 +21,68 @@ def initialize
...
+    letters = cells.map do |cell| 
+      cell.split('').first
+    end
+    #should we convert the letters to integers here instead of at the end?
+
+    numbers = cells.map do |cell|
+      cell.split('').last.to_i
+    end
+    # Could this be a case expression? 
+    # TDD by the books?
```
Comment (mkrumholz) [inline lib/board.rb:48]: I feel like this was resolved as well?

File: lib/board.rb [hunk 5]
```diff
@@ -21,8 +21,68 @@ def initialize
...
+    if ship.length != cells.length
```
Comment (mkrumholz) [inline lib/board.rb:49]: We can make this a helper method like Mike wanted :)

File: lib/board.rb [hunk 6]
```diff
@@ -21,8 +21,68 @@ def initialize
...
+      false 
+
+    elsif not_all_same?(letters) && not_all_same?(numbers)
```
Comment (mkrumholz) [inline lib/board.rb:52]: Should we make helper methods out of each of these conditional statements? Or is that excessive?

File: lib/board.rb (hunk 2 above)
Comment (suzkiee) [reply lib/board.rb:38]: great call!

File: lib/board.rb [hunk 7]
```diff
@@ -21,8 +21,68 @@ def initialize
...
+      false 
+
+    elsif all_same?(letters) && not_all_same?(numbers) 
+
+      numbers.each_cons(2).all? do |num_1, num_2|
```
Comment (mkrumholz) [inline lib/board.rb:57]: This definitely needs to be a helper method. Maybe we call it `consecutive?`

File: lib/board.rb (hunk 4 above)
Comment (suzkiee) [reply lib/board.rb:48]: Yes! These comments can go for sure!

File: lib/board.rb [hunk 8]
```diff
@@ -21,8 +21,68 @@ def initialize
...
+    # elsif letters.uniq.length > 1 && numbers.uniq.length == 1
+      #   # should these be helper methods
+      #   # Check letters to make sure they're consecutive
+      #   # Return true if letters are consecutive
+      # #   # False if not
+      # else
+      #   false 
+      # end
+    
+    else
```
Comment (mkrumholz) [inline lib/board.rb:70]: Whoops! Probably should clean this out/remove this 'else' statement for now

File: lib/board.rb (hunk 7 above)
Comment (suzkiee) [reply lib/board.rb:57]: I like that!

File: spec/board_spec.rb [hunk 9]
```diff
@@ -36,4 +36,21 @@
     end
   end
 
+  describe '#valid_placement?' do
+    
+    it 'checks if the placement is equal to length of ship' do
```
Comment (mkrumholz) [inline spec/board_spec.rb:41]: I feel like we should have a test for a case where this is true as well? Or just make sure other cases of this are accounted for?

File: lib/board.rb (hunk 6 above)
Comment (suzkiee) [reply lib/board.rb:52]: I think this is readable enough. I'm not sure what the convention is on nesting helpers inside helpers....

File: spec/board_spec.rb (hunk 9 above)
Comment (suzkiee) [reply spec/board_spec.rb:41]: Yeah, we'll make sure to add it in as we write tests for the rest of this method.

File: spec/board_spec.rb [hunk 10]
```diff
@@ -36,4 +36,21 @@
...
+      board = Board.new
+      cruiser = Ship.new("Cruiser", 3)
+
+      expect(board.valid_placement?(cruiser, ["A1", "A2"])).to be false
+    end
+
+    it 'checks if coordinates are consecutive' do
```
Comment (mkrumholz) [inline spec/board_spec.rb:48]: Maybe we should rename this to make sure it's clear this is the case where the ship is horizontal but the coords are not consecutive?

File: lib/board.rb (hunk 5 above)
Comment (mkrumholz) [reply lib/board.rb:70]: We decided to wait until the next round of refactoring for this one.

File: spec/board_spec.rb (hunk 10 above)
Comment (mkrumholz) [reply spec/board_spec.rb:48]: Will complete this in the next round of refactoring. It's on the list.

**Review by mkrumholz — COMMENTED** [review]
YAY! That made a HUGE difference to how organized I feel about this.

**Pull request merged**